### DISCUSSION TITLE
How to Dynamically Route and Authenticate Upstream Proxies in mitmproxy Based on URL?

### DISCUSSION THREAD

**Discussion by tranxuanthu** [mitmproxy/mitmproxy#6471] [Q&A]
I am working on a project using mitmproxy and I'm facing a challenge where I need to dynamically route requests to different upstream proxies based on the URL, along with handling authentication for these proxies. I would appreciate any guidance or suggestions on how to implement this.

Requirements:

Dynamic Proxy Routing:
- If the incoming request URL is https://example.com/123, mitmproxy should forward it through "Proxy A".
- If the URL is https://example.com/456, it should use "Proxy B".

Authentication for Each Proxy:
- Both "Proxy A" and "Proxy B" require authentication. The solution needs to handle this, ensuring the correct credentials are used based on which proxy is selected

Implementation in an Addon
- I am looking to implement this as an addon in mitmproxy, without using specific command-line arguments like --mode upstream:http://example.com:8081/.

My Attempts/Research: I've looked into the documentation but haven't found a clear way to change the upstream proxy dynamically based on the request URL, especially when it comes to incorporating authentication for different proxies.

Questions:

How can I programmatically route requests to different upstream proxies based on the URL in mitmproxy?
What is the most efficient method to authenticate with these proxies, keeping in mind that each proxy has different credentials?
Are there particular functions or modules within mitmproxy that I should look into for achieving this?
Any code examples, documentation references, or insights into how to approach this in mitmproxy would be extremely helpful.

Thank you in advance for your help!

below is the code I tried but not satisfied

import base64
from mitmproxy import http

class DynamicUpstreamProxy:
    def __init__(self):
        self.proxy_A = ("upstream-proxy-A.com", 8081)
        self.proxy_B = ("upstream-proxy-B.com", 8082)
        self.proxy_A_auth = self.encode_credentials("usernameA", "passwordA")
        self.proxy_B_auth = self.encode_credentials("usernameB", "passwordB")

    def encode_credentials(self, username, password):
        credentials = f"{username}:{password}"
        encoded_credentials = base64.b64encode(credentials.encode()).decode()
        return f"Basic {encoded_credentials}"

    def request(self, flow: http.HTTPFlow):
        url = flow.request.pretty_url

        if url.startswith("https://example.com/123"):
            # Upstream Proxy A
            flow.live.change_upstream_proxy_server(self.proxy_A)
            flow.request.headers["Proxy-Authorization"] = self.proxy_A_auth

        elif url.startswith("https://example.com/456"):
            #  Upstream Proxy B
            flow.live.change_upstream_proxy_server(self.proxy_B)
            flow.request.headers["Proxy-Authorization"] = self.proxy_B_auth

addons = [
    DynamicUpstreamProxy()
]

then run addon

mitmproxy -s my_upstream_addon.py

**Comment by seidnerj** [comment]
https://stackoverflow.com/questions/77450814/how-to-dynamically-route-and-authenticate-upstream-proxies-in-mitmproxy-based-on/77463155#77463155

**Reply by seidnerj** [reply]
Actually, the authentication piece with upstream proxy servers is not straightforward and not natively supported in the mitmproxy at the moment. One way to support something like would require patching mitmproxy.proxy.server.py in mitmxproxy, and adjusting the get_upstream_proxy method in the code found in the stackoverflow link you mentioned.  See here: https://github.com/mitmproxy/mitmproxy/pull/6504

**Comment by mhariri** [comment]
Just for future readers: I solved this by running two instances of mitmproxy, one that is just configured to handle upstreadm proxy auth and is configured using:
```
    cmd = [
        'mitmdump',
        '--listen-port', str(port),
        '--listen-host', '127.0.0.1',
        '--ssl-insecure',
        '--mode', f'upstream:{proxy_config["host"]}:{proxy_config["port"]}',
        '--upstream-auth', f'{proxy_config["username"]}:{proxy_config["password"]}',
        '--quiet'
    ]
    try:
                    process = subprocess.Popen(
                        cmd,
                        stdout=subprocess.PIPE,
                        stderr=subprocess.PIPE,
                        text=True
                    )
                    self.proxy_processes[proxy_id] = process
                    
                    logger.info(f"Started upstream proxy {proxy_id} on port {port} -> {proxy_config['host']}:{proxy_config['port']} (PID: {process.pid})")
                    
                except Exception as e:
                    logger.error(f"Failed to start upstream proxy {proxy_id}: {e}")
```

And then connecting to them using the pattern [here](https://raw.githubusercontent.com/mitmproxy/mitmproxy/refs/heads/main/examples/contrib/change_upstream_proxy.py)
